### ISSUE TITLE
Misbehaving of peer_allowed_ips

### ISSUE THREAD

**Issue by droscy** [droscy/esphome#2]
See comment https://github.com/esphome/esphome/pull/4256#issuecomment-1565547581

@thomas0bernard I've opened this issue to track your first problem, probably I need your help to perform tests.

**Comment by droscy** [comment]
So @thomas0bernard you have the malfunction from commit 7208401d to commit 1e338316.

Could you please do the following tests? Just to be sure of having understood correctly.

1. use wireguard from branch `fix/wrong_allowed_ips/1` with this configuration

   ```yaml
   external_components:
     - source:
         type: git
         url: https://github.com/droscy/esphome
         ref: fix/wrong_allowed_ips/1
       components:
         - wireguard
   
   wireguard:
     address: 10.44.100.100
     netmask: 255.255.0.0
     peer_endpoint: ***.com
     private_key: W***w=
     peer_public_key: 9***Y=
   ```

2. use wireguard from branch `fix/wrong_allowed_ips/2` with this configuration

   ```yaml
   external_components:
     - source:
         type: git
         url: https://github.com/droscy/esphome
         ref: fix/wrong_allowed_ips/2
       components:
         - wireguard
   
   wireguard:
     address: 10.44.100.100
     netmask: 255.255.0.0
     peer_endpoint: ***.com
     private_key: W***w=
     peer_public_key: 9***Y=
   ```

3. same as above but setting `netmask: 0.0.0.0`

Both tests 1 and 2 should work but you are reporting that 1 is working while 2 isn't. To fix issue in 2 you have to use 3. Correctly?

Which is the local network and IP of the esp board?

**Comment by thomas0bernard** [comment]
Thanks for the quick response @droscy . The requested tests reminded me that I was also using bgamari's fix (bgamari@ec905b6) to setup priorities, ensuring that the components are setup in this order: WiFi -> SNTP -> WG -> MQTT. Without it, MQTT over WG doesn't work (should we open another issue for that?). So I tested both of your branches with this commit merged into them.

> Which is the local network and IP of the esp board?

| interface | IP |
| ------ | ---------|
| **Local**               |192.168.159.134/24 |
| **WG**                  |10.44.100.100/16 |
| **MQTT broker** |10.0.1.10/16 |

> Both tests 1 and 2 should work but you are reporting that 1 is working while 2 isn't. To fix issue in 2 you have to use 3. Correctly?

Before your addition of `peer_allowed_ips`, the component was only routing packets destined to the same subnet (defined by address/netmask parameters) through WG. As my MQTT broker is in a different /16 subnet, **1.** should not work with 255.255.0.0, but should with 0.0.0.0, which is the case.

After your commit, if no `peer_allowed_ips` are defined, you add an allowed_ip of 0.0.0.0/0, as to route everything through WG by default. So in this case, both **2.** and **3.** should work (as should any netmask). **3.** works, **2.** does not, as if any other allowed_ip than the default one set from address/netmask is ignored. Same goes if I add anything in `peer_allowed_ips` (tried 0.0.0.0/0, 10.0.0.0/16, 10.0.1.10/32, all of the 3 should work, but none do).

**Comment by droscy** [comment]
In order to not make any mistake:
- 255.255.0.0 for netmask is the same as /16
- 10.0.1.10/16 and 10.44.100.100/16 are on different networks
- 10.0.0.0/8 covers both 10.0.1.0 and 10.44.100.100

> Before your addition of `peer_allowed_ips`, the component was only routing packets destined to the same subnet (defined by address/netmask parameters) through WG. As my MQTT broker is in a different /16 subnet, **1.** should not work with 255.255.0.0, but should with 0.0.0.0, which is the case.

Yes, you are right, it doesn't work: the netmask 255.255.0.0 prevents the system to use that interface for 10.0.1.10/16 even if wireguard will accept any traffic by default. If you set netmask to 0.0.0.0 your are telling the system to route there any traffic and then it works.

> After your commit, if no `peer_allowed_ips` are defined, you add an allowed_ip of 0.0.0.0/0, as to route everything through WG by default. So in this case, both **2.** and **3.** should work (as should any netmask).

No, it is not right, the same reason as above: if netmask is set to 255.255.0.0 the system will route through wireguard interface only 10.44.0.0/16, it will never use that interface for 10.0.0.0/16. So **2** should not work, while **3** should.

To make **2** work you need to set the netmask to 255.0.0.0 telling the system to route 10.0.0.0/8 through wireguard interface and set `peer_allowed_ips` either to `10.0.0.0/8` or to `[10.0.0.0/16, 10.44.0.0/16]`.

If my reasoning is right the following should work

```yaml
external_components:
  - source:
      type: git
      url: https://github.com/droscy/esphome
      ref: fix/wrong_allowed_ips/2  # also current wireguard/main
    components:
      - wireguard

wireguard:
  address: 10.44.100.100
  netmask: 255.0.0.0
  peer_endpoint: ***.com
  private_key: W***w=
  peer_public_key: 9***Y=
  peer_allowed_ips:
    - 10.0.0.0/16
    - 10.44.0.0/16
```

> The requested tests reminded me that I was also using bgamari's fix (https://github.com/bgamari/esphome/commit/ec905b6) to setup priorities, ensuring that the components are setup in this order: WiFi -> SNTP -> WG -> MQTT. Without it, MQTT over WG doesn't work (should we open another issue for that?)

I like that commit, really, but I think that it should be discussed after having wireguard merged in esphome because many components will be affected by it. Probably an official issue is a better place to discuss because it is about cross interaction between components, it is not directly related to wireguard itself. What you think?

**Comment by thomas0bernard** [comment]
I think I understand where our misunderstanding originates from. In my understanding of how I've been using WG on Linux, the address/subnet fields of the `[Interface]` section of a config file for wg-quick defines only the network interface to establish the WG connection, while `AllowedIPs` in the `[Peer]` section define what traffic should be routed on the interface, regardless of what the interface subnet is. 

That's what I was trying to reproduce with esphome and where my question came from, because I tested the same setup with a linux machine using the same peer keys (to validate the routing, at least), which worked succesfully with a 255.255.0.0 netmask on the interface address and an AllowedIP **outside** of that subnet.

If I understand correctly, your implementation uses the netmask as another limiting factor of what can be routed over the WG interface. If no `peer_allowed_ips` are specified, you allow everything in that subnet, whereas if there are specified `peer_allowed_ips` you allow only those **and** they have to be in the same subnet. Is this right?

> If my reasoning is right the following should work

I tested a few possibilities (using `netmask: 255.0.0.0`), and it seems like `peer_allowed_ips` has no effect. As expected, the default 0.0.0.0/0 works, as well as your example of 2 IPs above. Problem is, if I add only a single IP that's completely unrelated, if I understand correctly, I should not be able to connect to my MQTT broker. In this example, I still can:
```yaml
wireguard:
  address: 10.44.100.100
  netmask: 255.0.0.0
  peer_endpoint: ***.com
  private_key: W***w=
  peer_public_key: 9***Y=
  peer_allowed_ips:
    - 192.168.1.1/32
```

>I like that commit, really, but I think that it should be discussed after having wireguard merged in esphome because many components will be affected by it. Probably an official issue is a better place to discuss because it is about cross interaction between components, it is not directly related to wireguard itself. What you think?

Good idea opening another issue for that, let's have the conversation there

**Comment by droscy** [comment]
> If I understand correctly, your implementation uses the netmask as another limiting factor of what can be routed over the WG interface. If no peer_allowed_ips are specified, you allow everything in that subnet, whereas if there are specified peer_allowed_ips you allow only those and they have to be in the same subnet. Is this right?

Yes, it is. But it's not my decision, it is how lower [LwIP](https://savannah.nongnu.org/projects/lwip/) stack works on esp boards (as far as I understad, I'm not an expert of it).

Let me try to explain: on a Linux host the wireguard implementation creates an interaface (lets call it `wg0`) with netmask set to 255.255.255.255 then for each `AllowedIPs` a *static route* is added.

With this configuration

```bash
~# cat /etc/wireguard/wg0.conf
[Interface]
Address = 172.16.34.1

[Peer]
AllowedIPs = 172.16.34.3/32

[Peer]
AllowedIPs = 172.16.34.4/32

[Peer]
AllowedIPs = 172.16.34.5/32

[Peer]
AllowedIPs = 172.16.34.96/32
```

this is the created interface with its routing table

```bash
~# ip addr show dev wg0
124: wg0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none
    inet 172.16.34.1/32 scope global wg0
       valid_lft forever preferred_lft forever

~# ip route | grep wg0
172.16.34.3 dev wg0 scope link
172.16.34.4 dev wg0 scope link
172.16.34.5 dev wg0 scope link
172.16.34.96 dev wg0 scope link
```

In LwIP there is no way to create static routes as above (or I didn't find it) so the only way is to use the netmask **OR** to set the wireguard interface as the [system default](https://www.nongnu.org/lwip/2_0_x/group__netif.html#gac90f290edd005238d62aa94c4ac9dea3) (and it will be used when no route exists). I'm working on the second option but for now I will update the proposed documentation with a dedicated section for "netmask handling".

> I tested a few possibilities (using netmask: 255.0.0.0), and it seems like peer_allowed_ips has no effect. [...] I should not be able to connect to my MQTT broker. In this example, I still can:

Good catch, this can be a bug of my implementation! The real list used to allow (or drop) packets always includes `address/netmask` so if you setup esphome like this

```yaml
wireguard:
  address: 10.44.100.100
  netmask: 255.0.0.0
  peer_allowed_ips:
    - 192.168.1.1/32
```

the real list includes `10.44.100.100/8` and `192.168.1.1/32`, and that is why your test succedes. Originally I thought that the esp's own IP address should be always added to the list of allowed ips, otherwise one can drop its own packets. But, if netmask is not `255.255.255.255` we are implicitly allowing a whole network.

Probably it's better to only add `address/32` to the allowed list, leaving to the user the responsibility to set the appropriate allowed list in case of networks. I'll try to create a branch to have you test it. Thanks.

**Comment by thomas0bernard** [comment]
Thanks a lot for the detailed explanation. I've been working with the ESP32 for a few years now, but I've never dug under the hood of the networking implementation - happy to learn a bit more about it :smile: 

> so the only way is to use the netmask 

> But, if netmask is not 255.255.255.255 we are implicitly allowing a whole network.

Right, and correct me if I'm wrong, but if we are limited to use the subnet from address/netmask **and** that this whole subnet is automatically allowed, then the `peer_allowed_ips` parameter would have no purpose since:
- any IP **in** the subnet is already automatically allowed
- any IP **out** of the subnet can't be routed to

> but for now I will update the proposed documentation with a dedicated section for "netmask handling".

That's great, let me know if I can be of any help.

> Probably it's better to only add address/32 to the allowed list, leaving to the user the responsibility to set the appropriate allowed list in case of networks.

I think this solution would be perfect. No rush at all, I'm of course happy to help test it, and if there's anything I can do to help with the development, let me know.

**Comment by droscy** [comment]
> Right, and correct me if I'm wrong, but if we are limited to use the subnet from address/netmask and that this whole subnet is automatically allowed, then the peer_allowed_ips parameter would have no purpose since:

Yes, that's right.

> I think this solution would be perfect. No rush at all, I'm of course happy to help test it, and if there's anything I can do to help with the development, let me know.

@thomas0bernard could you please test `wireguard/dev` branch to see if it works correctly? I've set the default allowed ip to only `address/32` as discussed above. Thanks.

```yaml
external_components:
  - source:
      type: git
      url: https://github.com/droscy/esphome
      ref: wireguard/dev
    components:
      - wireguard
```

The relevant commit is https://github.com/droscy/esp_wireguard/commit/3d33f1458040da77814a165dca7d72e5d8c7a9a1

> [documentation] That's great, let me know if I can be of any help.

As soon as I push the changes I'll notify you so you can review the doc, many thanks.

**Comment by thomas0bernard** [comment]
> @thomas0bernard could you please test wireguard/dev branch to see if it works correctly?

Not quite sure why, but the change from `cv.positive_int` to `cv.int_range(min=1, max=65535)` in `__init__.py` broke config parsing for me:
```
Failed config

wireguard: [source wg.yaml:11]
  
  Don't know what '0' means as it has no time *unit*! Did you mean '0s'?.
  address: 10.44.100.100
  netmask: 255.0.0.0
  ...
```

Reverted that line and was able to test successfully. Everything works perfectly for me, thanks a lot!

**Comment by droscy** [comment]
> Not quite sure why, but the change from `cv.positive_int` to `cv.int_range(min=1, max=65535)` in `__init__.py` broke config parsing for me:

My mistake, the `min` value should be `0` not `1`. Fixed using `cv.uint16_t`.

Could you please test again, just to be sure. Thanks a lot.

**Comment by thomas0bernard** [comment]
> Could you please test again, just to be sure. Thanks a lot.

Works perfect, thanks!

**Comment by droscy** [comment]
Great, many thanks.

I've just released the version 0.3.0 of esp_wireguard and I'll merge the code.

Fixed with:
- https://github.com/droscy/esphome/commit/65df2ebd606de80168e0f404a45e38faa0439cc3
- https://github.com/droscy/esphome/commit/2cf01a972b952eca13b9765e10c701e12be929f4

@thomas0bernard here https://deploy-preview-2948--esphome.netlify.app/components/wireguard.html the updated documentation, have a look please.

**Comment by droscy** [comment]
Merged into `wireguard/main`
